pydantic/types.py

```diff
@@ -2429,3 +2437,92 @@ def __getattr__(self, item: str) -> Any:
                 return object.__getattribute__(self, item)
 
     __hash__ = object.__hash__
+
+
+_TAGGED_UNION_TAG_KEY = 'pydantic-tagged-union-tag'
+
+
+@_dataclasses.dataclass(**_internal_dataclass.slots_true, frozen=True)
```

Comment:
`frozen=True` is used here because annotations should be hashable, or you run into issues with caching generic containers (such as `typing.Union`!).

Technically we only need hashability, and we could get that with just `__hash__ = object.__hash__`, but looking at the fields/usage of both this and `CallableDiscriminator`, I didn't see any reason these types shouldn't just be frozen anyway. So that kept things simple.